client/connection.go

```diff
@@ -67,6 +67,43 @@ func ConnectLXD(url string, args *ConnectionArgs) (InstanceServer, error) {
 	return httpsLXD(url, args)
 }
 
+// ConnectVMAgent lets you connect to a VM agent over a VM socket.
+func ConnectVMAgent(vsockID int, args *ConnectionArgs) (InstanceServer, error) {
```

Comment:
@monstermunchkin what is this function for, I can't see it being used in the rest of the PR? I was surprised to see a VM specific piece of agent code in the client, as I had thought the client wouldn't change.

Reply:
This function is supposed to be used by LXD itself. However, I could move it to a separate PR. 
Example:
```go
func (vm *vmQemu) RenderState() (*api.InstanceState, error) {
	statusCode := vm.statusCode()
	pid, _ := vm.pid()

	agent, err := lxd.ConnectVMAgent(vm.vsockID(), &vm.agentConnectionArgs)
	if err == nil {
		status, _, err := agent.GetInstanceState("")
		if err == nil {
			status.Pid = int64(pid)
			status.Status = statusCode.String()
			status.StatusCode = statusCode

			return status, nil
		}
	}

	// At least return the Status and StatusCode if we couldn't get any
	// information for the VM agent.
	return &api.InstanceState{
		Pid:        int64(pid),
		Status:     statusCode.String(),
		StatusCode: statusCode,
	}, nil
}
```

Reply:
@monstermunchkin ah I see you're using the lxd client pkg to talk to the lxd agent running inside the VM.

Would it be possible to move the VM agent connect function out of the client package though, as my concern would be that we would be adding a vsock dependency to the public LXD client package?

Reply:
The `ConnectVMAgent` function itself doesn't depend on vsock. But it uses the `ProtocolLXD` struct and its unexported fields, and therefore should remain in the `client` package.
The `vsockHTTPClient` function in `client/util.go` does depend on vsock. But I don't think it makes sense to move a single function to a separate package. What package would you move it to?

A cleaner solution IMHO would be to use build tags for the client. That way, we could ignore all vsock functions if they are not needed.

Reply:
@monstermunchkin I see what you mean. On the one hand it is an "lxd client" but it just uses vsock to talk to the LXD process inside the VM, but on the other hand it is a publicly consumed package and so it would be prudent to limit its dependencies.

@stgraber what do you think?

Reply:
So I was thinking about this, couldn't we do something cleaner by having a ConnectLXDHTTP which takes in a http.Client?

That would let us move all the vsock logic into the lxd/ codebase.
We may even (though probably not now) get rid of the IsAgent() part entirely by having the passed in http.Client do the URL mangling for us.

Reply:
Done